v1/partial/uncompressed.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package partial
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"sync"
+
+	"github.com/google/go-containerregistry/v1"
+	"github.com/google/go-containerregistry/v1/types"
+	"github.com/google/go-containerregistry/v1/v1util"
+)
+
+// UncompressedImageCore represents the bare minimum interface a natively
+// uncompressed image must implement for us to produce a v1.Image
+type UncompressedImageCore interface {
+	imageCore
+
+	// UncompressedLayer is like UncompressedBlob, but takes the "diff id".
+	UncompressedLayer(v1.Hash) (io.ReadCloser, error)
+}
+
+// Assert that Image is a superset of this partial interface.
+var _ UncompressedImageCore = (v1.Image)(nil)
+
+// UncompressedToImage fills in the missing methods from an UncompressedImageCore so that it implements v1.Image.
+func UncompressedToImage(uic UncompressedImageCore) (v1.Image, error) {
+	return &uncompressedImageExtender{
+		UncompressedImageCore: uic,
+	}, nil
+}
+
+// uncompressedImageExtender implements v1.Image by extending UncompressedImageCore with the
+// appropriate methods computed from the minimal core.
+type uncompressedImageExtender struct {
+	UncompressedImageCore
```

Comment:
If you move these properties up to the UncompressedImage struct itself, could we avoid the transformation functions and embed the UncompressedImage struct directly in the sub-image type?

Something like:

```golang
// tarImage
type Image struct {
// tar stuff here
UncompressedImage
}
```

Reply:
Discussed a bit over chat last night, and it isn't clear there's a better way to do this.  Resolving for now, but open to later refactorings if we find something cleaner.